automation/orchestrator.go

```diff
@@ -0,0 +1,165 @@
+
+// RunWorkflow runs the given workflow
+func (o *Orchestrator) RunWorkflow(ctx context.Context, wid workflow.ID) error {
+	log.Debugf("RunWorkflow, workflow ID: %s", wid)
+	save := o.saveFactory(ctx)
+	wf, err := o.GetWorkflow(wid)
+	if err != nil {
+		log.Debugf("RunWorkflow: error getting workflow from store: %s", err)
+		return err
+	}
+	// need to replace w/ log collector
+	streams := ioes.NewDiscardIOStreams()
+
+	// TODO (ramfox): when hooks/completors are added, this should wait for the err, iterate through the hooks
+	// for this workflow, and emit the events for hooks that this orchestrator understands
+	runID := run.NewID()
+	return save(ctx, streams, wf, runID)
+}
+
+// DryRunWorkflow runs the given workflow, but does not record the output
+func (o *Orchestrator) DryRunWorkflow(ctx context.Context, wid workflow.ID) error {
+	log.Debugf("DryRunWorkflow, workflow ID: %s", wid)
+	apply := o.applyFactory(ctx)
+	wf, err := o.GetWorkflow(wid)
+	if err != nil {
+		log.Debugf("DryRunWorkflow: error getting workflow from store: %s", err)
+		return err
+	}
+	streams := ioes.NewDiscardIOStreams()
+
+	// TODO (ramfox): when we understand what it means to dryrun a hook, this should wait for the err, iterator thought the hooks
+	// for this workflow, and emit the events for hooks that this orchestrator understands
+	return apply(ctx, streams, wf)
+}
+
+// CreateWorkflow allows outside systems to create a workflow and save it to the workflow store using the orchestrator
+func (o *Orchestrator) CreateWorkflow(did string, pid profile.ID) (*workflow.State, error) {
+	// TODO (ramfox): when we add triggers in a follow up, this function should receive TriggerOptions as a param
+	// it should iterate over the triggers this orchestrator understands, and err if the workflow references
+	// any that it doesn't know about
+	// it should convert each TriggerOption into a Trigger & pass them down to `workflow.Create`
+	// TODO (ramfox): same goes for HookOptions & hooks
+	return o.workflows.Create(did, pid)
+}
+
+// GetWorkflow allows outside systems to get a workflow from the workflow store using the orchestrator
+func (o *Orchestrator) GetWorkflow(id workflow.ID) (*workflow.State, error) {
+	return o.workflows.Get(id)
+}
+
+// NewOrchestrator constructs an orchestrator, whose only responsibility, right now, is to create a workflow store & listen for trigger events
+func NewOrchestrator(ctx context.Context, bus event.Bus, saveFactory SaveFactory, applyFactory ApplyFactory, opts ...Option) (*Orchestrator, error) {
+	log.Debugf("NewOrchestrator opts=%v", opts)
+	ctx, cancel := context.WithCancel(ctx)
+	ok := false
+	orch := &Orchestrator{cancel: cancel}
+	defer func(o *Orchestrator) {
+		if !ok {
+			o.Shutdown()
+		}
+	}(orch)
+
+	o := &Options{}
+
+	for _, opt := range opts {
+		if err := opt(o); err != nil {
+			return nil, err
+		}
+	}
+
+	orch.bus = bus
+	orch.saveFactory = saveFactory
+	orch.applyFactory = applyFactory
+	orch.workflows = o.workflowStore
```

Comment:
process options before allocating the Orchestrator pointer & do assignments while allocating:

```go
orch := &Orchestrator{
  bus: bus,
  saveFactory: saveFactory,
  // ...
}
```